src/coreComponents/codingUtilities/Utilities.hpp

```diff
@@ -188,11 +188,14 @@ void forEachArgInTuple( std::tuple< Ts ... > const & tuple, F && func )
 
 template< typename VEC1, typename VEC2 >
 GEOSX_HOST_DEVICE
-void copy( integer const N, VEC1 const & v1, VEC2 const & v2 )
+void copy( integer const N,
```

Comment:
Sorry to bother you with this, but documenting why we need this `firstDeriv` could be helpful (in which context we need this, _e.g._)

More important, instead of providing `firstDeriv`, couldn't `v2` be a `range`/`slice` instead? That would keep this function more _mathematical_ and less _implementation aware_.

Reply:
Yes, as we discussed, it would be nice to passe a slice or range, but I don't know how to do it. 
Here for example:
https://github.com/GEOSX/GEOSX/blob/3e72aadf62b7083dd7f434052f3d90c8f3470230/src/coreComponents/constitutive/fluid/MultiFluidBase.hpp#L722
my `phaseFrac.derivs[ip]` is already a slice of size `numComp+2`, but I don't need first two entries (derivatives with respect to pressure and temperature). I would need a way to skip them, maybe by doing: 
```
applyChainRuleInPlace( numComp, dCompMoleFrac_dCompMassFrac, phaseFrac.derivs[ip][Deriv::dC], work );
```
Now that I am writing this, I should probably try that to see if it works.

For now, I added the Doxygen documentation in the `applyChainRule` function to explain why we have to pass `Deriv::dC`.

Reply:
In this case `v2` represents a general single-subscriptable entity, which may be a local array or an `ArraySlice`. The latter may not be contiguous in memory, and in that case precomputing an offset pointer is not safe or correct. So we have to rely on the type's `operator[]` to do the job at all times.

Theoretically it's possible to write a `SubSlice` type which represents a sub-selection of an `ArraySlice` along one or more dimensions. The type would store the offset(s) internally and apply as needed. But we don't have it yet.